starlette/datastructures.py

```diff
@@ -414,28 +413,25 @@ class UploadFile:
     An uploaded file included as part of the request data.
     """
 
-    spool_max_size = 1024 * 1024
     file: typing.BinaryIO
     headers: "Headers"
 
     def __init__(
         self,
         filename: str,
-        file: typing.Optional[typing.BinaryIO] = None,
+        file: typing.BinaryIO,
         content_type: str = "",
         *,
         headers: "typing.Optional[Headers]" = None,
```

Comment:
Given that we're changing the (internal-ish) interface at this point, should we switch it around so that `file` is the only required argument?...

```python
    def __init__(
         self,
         file: typing.Optional[typing.BinaryIO],
         *,
         filename: str = "",
         content_type: str = "",
         headers: "typing.Optional[Headers]" = None,
    ):
        self.file = file
        self.filename = filename or getattr(file, 'name', '')
        self.content_type = content_type
        self.headers = headers
```

Reply:
`UploadedFile.filename` was sanitized, what guarantee we provide that `file.name` also is?

Reply:
Yeah I think it would be best to just make it a constructor parameter and not fall back to `file.name`

Reply:
Okay, do we want `filename` and `content_type` *at all* here?

They're derived from the headers, so why not switch to...

```python
class UploadFile:
  def __init__(
         self,
         file: typing.Optional[typing.BinaryIO],
         *,
         headers: typing.Optional[Headers] = None,
    ):
        ...
```

Reply:
My vote would be to make `content_type` a property and keep filename.

The main reason for this is a practical one: extracting the filename from the headers requires knowledge that the request was originally a multipart/form-data request (that is, as far as I know, the only way a client would include that in the `content-disposition` header).
Also not every implementer of `BinaryIO` has a `.file` attribute (e.g. `BytesIO` does not).

Reply:
Okay - that's not unreasonable.

The one other practical case where I could see `UploadFile` being used is for raw binary uploads.

Eg. `PUT https://www.example.com/uploads/my-file-upload.txt`

In that kinda use-case it's useful having `filename` be something that can be provided explicitly, so that it can be derived from the URL path, rather than the HTTP headers.

So yeah, sounds good to me. 👍

Reply:
Yep that is a use case I personally really want: https://github.com/adriangb/xpresso/blob/c4455b29d323df1c59808ccafb306b9e7bbd80d1/xpresso/binders/_extractors/body/file.py#L40

I'm glad you brought it up and not me, I didn't want to introduce my biases 😅

Reply:
FYI I already pushed these changes / implementation @tomchristie, so I think this is ready for one last round of review for any typos/nits and then maybe a ✅

Reply:
Hi all. 

It would be nice if the release could include the removal of `content_type` from the UploadFile, as this is a breaking change.